File: Editorial/22_Generate_Parentheses/Go/3_Divide_and_Conquer.go
func generateParenthesis(n int) []string {
    if n == 0 {
        return []string{""}
    }
    var result []string
    for leftCount := 0; leftCount < n; leftCount++ {
        leftStrings := generateParenthesis(leftCount)
        rightStrings := generateParenthesis(n - 1 - leftCount)
        for _, leftString := range leftStrings {
            for _, rightString := range rightStrings {
                result = append(result, "("+leftString+")"+rightString)
            }
        }
    }
    return result
}

File: Editorial/0022_Generate_Parentheses/CPP/3_Divide_and_Conquer.cpp
class Solution {
public:
    std::vector<std::string> generateParenthesis(int n) {
        if (n == 0) {
            return std::vector<std::string>{""};
        }
        std::vector<std::string> answer;
        for (int leftCount = 0; leftCount < n; ++leftCount) {
            auto leftStrings = generateParenthesis(leftCount);
            auto rightStrings = generateParenthesis(n - 1 - leftCount);
            for (const auto& leftString : leftStrings) {
                for (const auto& rightString : rightStrings) {
                    answer.push_back("(" + leftString + ")" + rightString);
                }
            }
        }
        return answer;
    }
};
